src/script/lua_api/l_metadata.cpp

```diff
@@ -166,25 +170,50 @@ int MetaDataRef::l_get_float(lua_State *L)
 
 	std::string str_;
 	const std::string &str = meta->getString(name, &str_);
-	// Convert with Lua, as is done in set_float.
 	lua_pushlstring(L, str.data(), str.size());
-	lua_pushnumber(L, lua_tonumber(L, -1));
+	f64 number;
+	// We can't be sure that Lua's `tonumber` handles nan / (-)inf properly
+	if (str == "nan") {
+		number = std::numeric_limits<f64>::quiet_NaN();;
+	} else if (str == "inf") {
+		number = std::numeric_limits<f64>::infinity();
+	} else if (str == "-inf") {
+		number = -std::numeric_limits<f64>::infinity();
+	} else {
+		number = lua_tonumber(L, -1);
+	}
+	lua_pushnumber(L, number);
 	return 1;
 }
 
+static inline std::string double_to_string(f64 number)
+{
+	// Do not use Lua to convert number to string, since that may lose precision.
+	if (std::isfinite(number)) {
+		char buf[64];
+		snprintf(buf, sizeof(buf), "%.17g", number);
+		return buf;
+	}
+	// We can't be sure of how %g formats these, so do it ourselves.
+	if (number < 0)
+		return "-inf";
+	if (number > 0)
+		return "inf";
+	return "nan";
+
+}
```

Comment:
would make sense to move these to a header
(along with a `void lua_string_to_double(lua_State *L, std::string_view str)`)

Reply:
done, put it in `LuaHelper`. i'm not 100% happy with that (among other reasons because i don't like abusing inheritance to get such helpers into scope) but it should be okay now.